Frameworks/include/CGImageInternal.h

```diff
@@ -34,93 +40,209 @@ class DisplayTexture;
-    CGImageRef _parent;
+  CGImageRef _parent;
+
+  virtual ~CGImageBacking() {}
+
+  virtual CGImageRef Copy() = 0;
+
+  virtual void GetPixel(int x, int y, float &r, float &g, float &b,
+                        float &a) = 0;
+  virtual int InternalWidth() = 0;
+  virtual int InternalHeight() = 0;
+  virtual int Width() = 0;
+  virtual int Height() = 0;
+  virtual int BytesPerRow() = 0;
+  virtual int BytesPerPixel() = 0;
+  virtual int BitsPerComponent() = 0;
+  virtual ID2D1RenderTarget *GetRenderTarget() = 0;
+  virtual void GetSurfaceInfoWithoutPixelPtr(__CGSurfaceInfo *surfaceInfo) = 0;
+  virtual __CGSurfaceFormat SurfaceFormat() = 0;
+  virtual CGColorSpaceModel ColorSpaceModel() = 0;
+  virtual CGBitmapInfo BitmapInfo() = 0;
+  virtual void *StaticImageData() = 0;
+  virtual void *LockImageData() = 0;
+  virtual void ReleaseImageData() = 0;
+  virtual cairo_surface_t *LockCairoSurface() = 0;
+  virtual void ReleaseCairoSurface() = 0;
+  virtual void SetFreeWhenDone(bool freeWhenDone) = 0;
+  virtual DisplayTexture *GetDisplayTexture() { return NULL; }
+
+  virtual void DiscardIfPossible() {}
+  virtual void ConstructIfNeeded() {}
+  virtual bool DrawDirectlyToContext(CGContextImpl *ctx, CGRect src,
+                                     CGRect dest) {
+    return false;
+  }
+  virtual CGImageRef CopyOnWrite();
+};
 
-    virtual ~CGImageBacking() {
-    }
+typedef enum {
+  CGImageTypeBitmap,
+  CGImageTypeGraphicBuffer,
+  CGImageTypeNoData,
+  CGImageTypeOpenGL,
+  CGImageTypeVector,
+  CGImageTypePNG,
+  CGImageTypeJPEG
+} CGImageType;
 
-    virtual CGImageRef Copy() = 0;
-
-    virtual void GetPixel(int x, int y, float& r, float& g, float& b, float& a) = 0;
-    virtual int InternalWidth() = 0;
-    virtual int InternalHeight() = 0;
-    virtual int Width() = 0;
-    virtual int Height() = 0;
-    virtual int BytesPerRow() = 0;
-    virtual int BytesPerPixel() = 0;
-    virtual int BitsPerComponent() = 0;
-    virtual ID2D1RenderTarget* GetRenderTarget() = 0;
-    virtual void GetSurfaceInfoWithoutPixelPtr(__CGSurfaceInfo* surfaceInfo) = 0;
-    virtual __CGSurfaceFormat SurfaceFormat() = 0;
-    virtual CGColorSpaceModel ColorSpaceModel() = 0;
-    virtual CGBitmapInfo BitmapInfo() = 0;
-    virtual void* StaticImageData() = 0;
-    virtual void* LockImageData() = 0;
-    virtual void ReleaseImageData() = 0;
-    virtual cairo_surface_t* LockCairoSurface() = 0;
-    virtual void ReleaseCairoSurface() = 0;
-    virtual void SetFreeWhenDone(bool freeWhenDone) = 0;
-    virtual DisplayTexture* GetDisplayTexture() {
-        return NULL;
+// TODO: use for reverse mapping of GetPixelFormat
+typedef struct {
+  CGColorSpaceModel colorSpaceModel;
+  CGBitmapInfo bitmapInfo;
+  unsigned int bitsPerComponent;
+  unsigned int bytesPerPixel;
+} __CGImagePixelProperties;
+
+struct __CGImageImpl {
+  CFRuntimeBase _base;
+  Microsoft::WRL::ComPtr<IWICBitmapSource> bitmapImageSource{nullptr};
+  CGBitmapInfo bitmapInfo;
+  bool isMask;
```

Comment:
>isMask [](start = 7, length = 6)

The scalar types, you do need to initialize, this probably warrants a constructor.